signature-crate/src/sign/digest.rs

```diff
@@ -14,5 +18,47 @@ where
     S: Signature,
 {
     /// Sign the given prehashed message `Digest`, returning a signature.
-    fn sign(&self, digest: D) -> Result<S, Error>;
+    fn sign_digest(&self, digest: D) -> Result<S, Error>;
+}
+
+/// Sign the given message using a "raw digest" signature algorithm, i.e.
+/// any algorithm where signatures are always computed as `S(H(m)))` where:
+///
+/// - `S`: signature algorithm
+/// - `H`: hash (a.k.a. digest) function
+/// - `m`: message
+///
+/// This is the preferred trait to be `impl`'d for such algorithms, and when
+/// used will take advantage of a blanket `impl` of `Sign` which will hash
+/// the original message in advance using the relevant `Digest` algorithm
+pub trait SignRawDigest<S>: Send + Sync
+where
+    S: Signature + RawDigestSignature,
+{
+    /// Digest algorithm to hash the input message with
+    type Digest: Digest;
+
+    /// Sign the given prehashed message `Digest`, returning a signature.
+    fn sign_raw_digest(&self, digest: Self::Digest) -> Result<S, Error>;
+}
+
+impl<D, S, T> SignDigest<D, S> for T
+where
+    D: Digest,
+    S: Signature + RawDigestSignature,
+    T: SignRawDigest<S, Digest = D>,
```

Comment:
@newpavlov here is where the `Digest` associated type from `SignRawDigest` is used. It's needed to constrain the blanket impl.